ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+        testAdhoc.pay(1234, 1234);
+        testExit.ticketInserted(testAdhoc.getBarcode());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");
+    }
+    
+    @Test
+    public void testAlternateFlowAdhoc() {          //test accurate detection of rejected adhoc ticket
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        assertEquals(testExit.getState(), "WAITING");
+        
+        when(mockAdhocDAO.findTicketByBarcode("Atest")).thenReturn(null);
+        testAdhoc.pay(1234, 1234);
+        testExit.ticketInserted(testAdhoc.getBarcode());
+        assertEquals(testExit.getState(), "REJECTED");
+    }
+    
+    @Test
+    public void testAlternateFlowSeason() {          //test accurate detection of rejected season ticket
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        assertEquals(testExit.getState(), "WAITING");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "REJECTED");
+    }
+    
+    @Test(expected=RuntimeException.class)
+    public void testAlternateFlowNullSeason() {          //test null season ticket
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        assertEquals(testExit.getState(), "WAITING");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(null);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "REJECTED");
+    }
+    
+    @Test
+    public void testAlternateFlowNullAdhoc() {          //test null adhoc ticket
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        assertEquals(testExit.getState(), "WAITING");
+        
+        when(mockAdhocDAO.findTicketByBarcode(testSeason.getId())).thenReturn(null);
+        testExit.ticketInserted(testAdhoc.getBarcode());
+        assertEquals(testExit.getState(), "REJECTED");
+    }
+}

```

Comment:
another test that probably should be added is what happens if someone drives up to the exit gate and then backs up and decides to stay.
If you check the activity diagram it is there

Reply:
Added a test for PROCESSED -> BLOCKED.